### PR TITLE
Convert Project To Relase Relations to Enums

### LINKED ISSUE #257: enum for project-release relation?
It feels strange that we have enums for the project-project relationship and for the release-release-relationship, but not for the project-release-relationship. The latter is described by a string. 

Do we need more structure here?

This would of course mean that we have to migrate existing data...

### PR DESCRIPTION
**Pull request by alex-evo**
- convert project to release relationship description from string to a structure with a `ReleaseRelationship` and a `MainlineState`
- display `MainlineState`s of project-release-relations on Project view/Clearing State page.

closes #257

### CODE REVIEW

**Comment by mcjaeger** [comment]
I appreciate the early access, just in case you have not seen it, it came during testing:
```
An error occurred at line: 112 in the jsp file: /html/components/includes/releases/editReleaseInformation.jspf
MainlineState cannot be resolved to a type
109:                     name="<portlet:namespace/><%=Release._Fields.MAINLINE_STATE%>"
110:                     style="min-width: 162px; min-height: 28px;">
111: 
112:                 <sw360:DisplayEnumOptions type="<%=MainlineState.class%>" selected="${release.mainlineState}"/>
113:             </select>
114:         </td>
115: 
` ``

**Comment by alex-evo** [comment]
One minor UI problem is documented in #409 and will be fixed separately.

File: backend/src-common/src/main/java/org/eclipse/sw360/datahandler/entitlement/ReleaseModerator.java [hunk 1]
```diff
@@ -10,6 +10,7 @@
 
 import org.eclipse.sw360.datahandler.common.Moderator;
 import org.eclipse.sw360.datahandler.common.SW360Utils;
+import org.eclipse.sw360.datahandler.thrift.ReleaseRelationship;
```
Comment (maximilianhuber) [inline backend/src-common/src/main/java/org/eclipse/sw360/datahandler/entitlement/ReleaseModerator.java]: Unnecessary?

Comment (alex-evo) [inline backend/src-common/src/main/java/org/eclipse/sw360/datahandler/entitlement/ReleaseModerator.java]: No. Used at https://github.com/sw360/sw360portal/pull/395/files/49c7da2a83d3922cee6c0d8765bb9ebdf4c0471d#diff-5ae0339c1ee942a32403fc4eaef8aadaR103
Used to be imported by line https://github.com/sw360/sw360portal/pull/395/files/49c7da2a83d3922cee6c0d8765bb9ebdf4c0471d#diff-5ae0339c1ee942a32403fc4eaef8aadaR16

File: libraries/lib-datahandler/src/main/thrift/projects.thrift [hunk 2]
```diff
@@ -93,7 +96,7 @@ struct Project {
 
     // Linked objects
     30: optional map<string, ProjectRelationship> linkedProjects,  //id, relationship
-    31: optional map<string, string> releaseIdToUsage,    //id, comment
+    31: optional map<string, ProjectReleaseRelationship> releaseIdToUsage,    //id, comment
```
Comment (maximilianhuber) [inline libraries/lib-datahandler/src/main/thrift/projects.thrift]: The comment is out of date. Remove it?

File: backend/src/src-moderation/src/main/java/org/eclipse/sw360/moderation/db/ModerationRequestGenerator.java [hunk 3]
```diff
@@ -113,30 +116,28 @@ protected void dealWithStringsEnumsStructs(U field) {
...
         }
 
         //determine changes in common linkedProjects
         Set<String> commonKeys = Sets.intersection(updateMap.keySet(), actualMap.keySet());
         for(String id : commonKeys) {
-            String actual = actualMap.get(id);
-            String update = updateMap.get(id);
+            Object actual = actualMap.get(id);
+            Object update = updateMap.get(id);
             if(! actual.equals(update)) {
```
Comment (alex-evo) [inline backend/src/src-moderation/src/main/java/org/eclipse/sw360/moderation/db/ModerationRequestGenerator.java]: No, it uses the overridden method of whatever object is referenced

File: frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/users/UserCacheHolder.java [hunk 4]
```diff
@@ -45,6 +49,12 @@ private User getCurrentUser(PortletRequest request) {
 
         return loadUserFromEmail(email);
     }
+    private String getCurrentUserEmail(ServletRequest request) {
+        ThemeDisplay themeDisplay = (ThemeDisplay)request.getAttribute(WebKeys.THEME_DISPLAY);
+
+        com.liferay.portal.model.User liferayUser = themeDisplay.getUser();
+        return liferayUser == null ? null : liferayUser.getEmailAddress();
```
Comment (alex-evo) [inline frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/users/UserCacheHolder.java]: good point

File: frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/tags/DisplayProjectChanges.java [hunk 5]
```diff
@@ -263,7 +267,7 @@ private void renderReleaseIdToUsage(StringBuilder display) {
                    additions.getReleaseIdToUsage().keySet(),
                    changedReleaseIds);
 
-           LinkedReleaseRenderer renderer = new LinkedReleaseRenderer(display, tableClasses, idPrefix, actual.getCreatedBy());
+           LinkedReleaseRenderer renderer = new LinkedReleaseRenderer(display, tableClasses, idPrefix, userEmail);
```
Comment (alex-evo) [inline frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/tags/DisplayProjectChanges.java]: This email address is used to load the user on behalf of whom the releases are requested in `LinkedReleaseRenderer`. Two problems with that:
- I don't think current moderator should load releases on behalf of someone else.
- The old code throws an error if the user named in `createdBy` does not exist in the system. Which can happen when you import data from a legacy application, for example. The consequence in that case, is that one cannot review and accept or decline the moderation request at all.

File: frontend/sw360-portlet/src/main/webapp/html/projects/includes/projects/clearingStatus.jspf [hunk 6]
```diff
@@ -7,13 +6,14 @@
...
 --%>
+<%@ page import="org.eclipse.sw360.portal.common.PortalConstants" %>
 <portlet:resourceURL var="deleteReleaseAjaxURL">
     <portlet:param name="<%=PortalConstants.ACTION%>" value='<%=PortalConstants.DELETE_RELEASE%>'/>
 </portlet:resourceURL>
 
 <core_rt:set var="clearingTeamsStringSet" value='<%=PortalConstants.SET_CLEARING_TEAMS_STRING%>'/>
 
-<jsp:useBean id="releasesAndProjects" type="java.util.Map<org.eclipse.sw360.datahandler.thrift.components.Release,java.lang.String>" scope="request"/>
+<jsp:useBean id="releasesAndProjects" type="java.util.Map<org.eclipse.sw360.datahandler.thrift.components.Release,java.lang.String[]>" scope="request"/>
```
Comment (alex-evo) [inline frontend/sw360-portlet/src/main/webapp/html/projects/includes/projects/clearingStatus.jspf]: That's a quick and dirty solution. You're absolutely right: I should've cleaned that up before creating PR

File: frontend/sw360-portlet/src/main/webapp/html/projects/includes/projects/clearingStatus.jspf [hunk 7]
```diff
@@ -71,8 +73,9 @@
             columns: [
                 {title: "Name"},
                 {title: "Project Origin"},
-                {title: "Release Clearing State"},
-                {title: "Release Mainline State"},
+                {title: "Clearing State"},
+                {title: "Mainline State"},
+                {title: "Project Mainline State"},
```
Comment (alex-evo) [inline frontend/sw360-portlet/src/main/webapp/html/projects/includes/projects/clearingStatus.jspf]: The change is:
`Release Clearing State` -> `Clearing State`
`Release Mainline State` -> `Mainline State`
new column `Project Mainline State`

File: libraries/lib-datahandler/src/main/thrift/projects.thrift (hunk 2 above)
Comment (alex-evo) [inline libraries/lib-datahandler/src/main/thrift/projects.thrift]: yes

**Comment by alex-evo** [comment]
Corrected and rebased. @maximilianhuber, please have a look

**Comment by maximilianhuber** [comment]
@alex-evo is the issue in comment https://github.com/sw360/sw360portal/pull/395#issuecomment-300917236 fixed?

**Comment by alex-evo** [comment]
Yes, of course. It was an artifact of a WIP state

**Comment by mcjaeger** [comment]
compilations and things work, I am on it to test the datamodel

**Comment by maximilianhuber** [comment]
@alex-evo while @mcjaeger is testing you could start by merging the other PRs?

**Pull request merged**